backends/dpdk/dpdkArch.cpp

```diff
@@ -772,26 +948,112 @@ const IR::Node *ConvertBinaryOperationTo2Params::postorder(IR::P4Parser *a) {
+     * a new field of the same header type in the main header structure.
+     * If the main header structure is placed before the definition of that
+     * header, it is a problem for Type Checking.
+     * There might also be header definitions internally created by compiler
+     * which would cause the same problem.
+     * Therefore here we move all structure and header definitons directly
+     * before the main header structure to prevent this problem.
+     */
+
+    /* Find main header structure. */
+    IR::Vector<IR::Node>::iterator headerTypeIter = std::find_if(p->objects.begin(),
+            p->objects.end(), [this](const IR::Node *n){
+                return n->is<IR::Type_Struct>() &&
+                       n->to<IR::Type_Struct>()->name == structure->header_type;});
+    auto mainHeaderStructure = *headerTypeIter;
+
+    /**
+     * Store header definitions which will be moved before the main header structure
+     * into temporary vector and remove them temporarily from the objects vector.
+     */
+    IR::IndexedVector<IR::Node> defs;
+    std::copy_if(headerTypeIter, p->objects.end(), std::back_inserter(defs),
+            [this](const IR::Node *n){
+                return n->is<IR::Type_StructLike>() &&
+                n->to<IR::Type_StructLike>()->name != structure->header_type;});
+    p->objects.erase(std::remove_if(headerTypeIter, p->objects.end(),
+            [this](const IR::Node *n){
+                return n->is<IR::Type_StructLike>() &&
+                n->to<IR::Type_StructLike>()->name != structure->header_type;}), p->objects.end());
+
+    /* Insert header definitions before the main header structure. */
+    p->objects.insert(headerTypeIter, defs.begin(), defs.end());
+
+    LOG2("Structure and header definitions:");
+    for (auto d : defs) {
+        LOG2(" " << d);
+    }
+    LOG2("moved before main header structure:" << std::endl << " " << mainHeaderStructure);
+
     return p;
 }
 
-const IR::Node *CollectLocalVariableToMetadata::postorder(IR::Type_Struct *s) {
+const IR::Node *CollectLocalVariables::postorder(IR::Type_Struct *s) {
     if (s->name.name == structure->local_metadata_type) {
         for (auto kv : locals_map) {
             for (auto d : kv.second) {
                 if (auto dv = d->to<IR::Declaration_Variable>()) {
-                    s->fields.push_back(new IR::StructField(
-                        IR::ID(kv.first + "_" + dv->name.name), dv->type));
+                    auto type = typeMap->getType(dv, true);
+                    if (type->is<IR::Type_Header>()) {
+                        LOG3("Variable: " << dv << std::endl <<
+                             " type: " << type << std::endl <<
+                             " already added to: " << structure->header_type);
+                    } else {
+                        auto sf = new IR::StructField(
+                            IR::ID(kv.first + "_" + dv->name.name), dv->type);
```

Comment:
it's not obvious to me that this field name could not clash with another field name of the same struct.

Reply:
In fact it was already done this way before. I think it relies on the fact that it is appending control/parser block name which has to be unique and local variable name which has to be unique in context of its control/parser block. But I think it could be theoretically possible to have the same name in the original P4 source code which could clash with this newly created name. I'm rather adding refMap->newName() here.

Reply:
So it's not that simple. Same name, as here, needs to be used also in `const IR::Node *CollectLocalVariables::postorder(IR::PathExpression *p)`.
That would mean that the name would have to be generated in:
```
const IR::Node *CollectLocalVariables::preorder(IR::P4Program *p) {
    for (auto kv : structure->parsers) {
        locals_map.emplace(kv.first+"_parser", kv.second->parserLocals);
    }
    for (auto kv : structure->pipelines) {
        locals_map.emplace(kv.first, kv.second->controlLocals);
    }
    for (auto kv : structure->deparsers) {
        locals_map.emplace(kv.first+"_deparser", kv.second->controlLocals);
    }
```
which would require iterating over all controlLocals and creating the new names.

So I suggest to leave it as it is right now.

What do you think?

Reply:
Many of our analyses have two passes: an Inspector which generates information (e.g., new names), and a Transform which makes the changes according to the generated information. This probably won't require much work - just moving some code around and storing results in a HashMap. I would say that safety is the most important goal, so you really should make sure names are unique.

Reply:
I suggest to merge this change in separate PR, as there is a diff in quite a lot of spec files, so it would pollute this PR with unrelated changes. I prepared another PR with these changes based on this PR: https://github.com/p4lang/p4c/pull/3080, which could be merged when this PR is merged. What do you think?